### PR TITLE
Add more Kibana verification in upgrade

### PR DESCRIPTION
**Pull request by anpingli**
Check two Kibana users during upgrade. One is authorized via group. (BZ1895581)
Move some steps to step definition and use the step 'custerlogging to be functional in the project'
Check new project log instead of comparing the doc number. I am worried about the Noisy data from buffer.
persistent some projects during upgrade.

### CODE REVIEW

**Comment by anpingli** [comment]
@QiaolingTang  Could you review the upgrade changes?
-  Check two Kibana users during upgrade. One is authorized via group. (BZ1895581)
-  Move some steps to step definition and use the step 'custerlogging to be functional in the project'
-  Check new project log instead of comparing the doc number. I am worried about the Noisy data from buffer. 
-  persistent  some projects during upgrade.

File: features/step_definitions/logging.rb [hunk 1]
```diff
@@ -1056,3 +1056,125 @@
...
+
+Given /^I have clusterlogging with(?: (\d+))? persistent storage ES$/ do |es_num|
+    # I switch to cluster admin pseudo user
+    ensure_admin_tagged
+    es_num ||= 3
+    redundancy_policy="SingleRedundancy"
+    deploy_cluster_logging=true
+
+    @result = admin.cli_exec(:get, resource: 'clusterlogging', resource_name: "instance", o: "name")
+    if @result[:exitstatus] == 0
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1160]: @anpingli  `if cluster_logging('instance').exists? ` should work here.

File: features/step_definitions/logging.rb [hunk 2]
```diff
@@ -1056,3 +1056,125 @@
     | pod_label       | #{pod_label}     |
   })
 end
+
+
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1060]: Too many new lines

File: features/step_definitions/logging.rb [hunk 3]
```diff
@@ -1056,3 +1056,125 @@
...
+        step %Q/I run the :new_project client command with:/,table(%{
+          | project_name | #{project_name} |
+        })
+        raise "Error creating namespace" unless @result[:success]
+    end
+
+    case log_type
+      when "json"
+        template_file = "#{file_dir}/container_json_log_template.json"
+      when "flat"
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1076]: @anpingli  what's the difference between `flat` and `unicode`? From the file name, I see now difference.

File: features/step_definitions/logging.rb [hunk 4]
```diff
@@ -1056,3 +1056,125 @@
...
+    unless project(project_name).exists?
+        step %Q/I run the :new_project client command with:/,table(%{
+          | project_name | #{project_name} |
+        })
+        raise "Error creating namespace" unless @result[:success]
+    end
+
+    case log_type
+      when "json"
+        template_file = "#{file_dir}/container_json_log_template.json"
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1075]: From the whole `case` block, I think no need to declare `json` here.

File: features/step_definitions/logging.rb (hunk 3 above)
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1076]: If you need to declare these 2 types, then you can simplify it to:
```
      when "unicode", "flat"
        template_file = "#{file_dir}/container_json_unicode_log_template.json"
```

File: features/step_definitions/logging.rb [hunk 5]
```diff
@@ -1056,3 +1056,125 @@
...
+      when "unicode"
+        template_file = "#{file_dir}/container_json_unicode_log_template.json"
+      else
+        template_file = "#{file_dir}/container_json_log_template.json"
+      end
+
+    unless  project.pods(by:user).size > 0 
+        step %Q/I run the :new_app client command with:/,table(%{
+           | file | #{template_file} |
+         })
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1087]: @anpingli Do we need to check the pod status here?

File: features/step_definitions/logging.rb [hunk 6]
```diff
@@ -1056,3 +1056,125 @@
...
+            | index_pattern_name | "#{pattern_name}" |
+         })
+         raise "Failed to create index pattern #{pattern_name}" unless @result[:success]
+    end
+end
+
+Given /^cluster-admin create the #{QUOTED} pattern for the #{WORD} user$/ do | pattern_name, who |
+    user(word_to_num(who))
+    user_name=user.name
+    user_token=user.cached_tokens.first
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1107]: @anpingli Should these lines under `step %Q/I switch to cluster admin pseudo user/` ?

File: features/step_definitions/logging.rb [hunk 7]
```diff
@@ -1056,3 +1056,125 @@
...
+    unless @result[:success]
+        step %Q/I run the :go_to_kibana_management_page web action/
+        step %Q/I perform the :create_index_pattern_in_kibana web action with:/, table(%{
+            | index_pattern_name | "#{pattern_name}" |
+         })
+         raise "Failed to create index pattern #{pattern_name}" unless @result[:success]
+    end
+end
+
+Given /^cluster-admin create the #{QUOTED} pattern for the #{WORD} user$/ do | pattern_name, who |
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1104]: @anpingli Could you please update the step name? `cluster-admin create xxxx` can easily make some misunderstanding here.

File: features/step_definitions/logging.rb [hunk 8]
```diff
@@ -1056,3 +1056,125 @@
...
+    })
+    raise("Can not select to pattern #{ pattern_name}!") unless @result[:success]
+
+    step %Q/I perform the :search_doc_in_kibana web action with:/,table(%{
+             | search_string | #{lucene_query_string} |
+    })
+    raise("Can not refresh the search #{lucene_query_string}!") unless @result[:success]
+
+    retries = 3
+    while true
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1138]: @anpingli I think here the max trier is 4. And using `while true` is too dangerous.

Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1138]:
```
    retries = 3
    while true
        puts retries
        if retries > 0
           puts("Can not find documents using clause under pattern ! retry.....")
        else
           raise("Can not find documents using clause under pattern ! after 3 retry,  abort")
        end
        retries -= 1
    end
```
executed above block got below output:
```
3
Can not find documents using clause under pattern ! retry.....
2
Can not find documents using clause under pattern ! retry.....
1
Can not find documents using clause under pattern ! retry.....
0
test.rb:7:in `<main>': Can not find documents using clause under pattern ! after 3 retry,  abort (RuntimeError)
```

File: features/step_definitions/logging.rb [hunk 9]
```diff
@@ -1056,3 +1056,125 @@
...
+           log("Can not find documents using clause #{lucene_query_string} under pattern  #{ pattern_name}! retry.....")
+        else
+           raise("Can not find documents using clause #{lucene_query_string} under pattern  #{ pattern_name}! after 3 retry,  abort")
+        end
+        step %Q/I run the :click_refresh_button_in_kibana action/
+        retries -= 1
+    end
+end
+
+Given /^I have clusterlogging with(?: (\d+))? persistent storage ES$/ do |es_num|
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1152]: @anpingli  why do you add this step here? The step `I create clusterlogging instance with` doesn't work for you?

File: features/upgrade/logging/upgrade.feature [hunk 10]
```diff
@@ -6,88 +6,80 @@ Feature: Logging upgrading related features
...
+    Then evaluation of `project.name` is stored in the :proj1 clipboard
+    And I have "json" log pod in project "<%= cb.proj1 %>"
+    Given I wait for clusterlogging to be functional in the project
     # check the logs collected before upgrading
-    Given I wait for the project "logging-upgrade-data-check" logs to appear in the ES pod
-    # check if logging stack could gather logs
-    And I wait for the project "<%= cb.proj.name %>" logs to appear in the ES pod
-    And evaluation of `cb.doc_count` is stored in the :docs_count_1 clipboard
+    Given I switch to cluster admin pseudo user
+    Given I use the "openshift-logging" project
```
Comment (QiaolingTang) [inline features/upgrade/logging/upgrade.feature:49]: @anpingli No need to add these 2 steps, they're executed in `I wait for clusterlogging to be functional in the project`

File: features/step_definitions/logging.rb (hunk 4 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1075]: I'd like keep the 'json' declaration.  That make the function more specific.

File: lib/rules/web/admin_console/4.9/logging_kibana.xyaml [hunk 11]
```diff
@@ -195,3 +196,37 @@ click_visuallize_button_in_discover_page:
...
+    timeout: 30
+  - selector:
+      xpath: //div[@id="indexPatternListReact"]//*[text()="<index_pattern_name>"]
+
+select_index_pattern_in_kibana:
+  action: go_to_kibana_discover_page
+  action:
+    if_element:
+      selector:
+        xpath: //span[@class="euiSideNavItemButton__label" and contains(., "Index Patterns")]
```
Comment (QiaolingTang) [inline lib/rules/web/admin_console/4.9/logging_kibana.xyaml:214]: @anpingli I can't find this element in `Discover` page, from line 203, I think it should be in the `Management` page.

File: lib/rules/web/admin_console/4.9/logging_kibana.xyaml [hunk 12]
```diff
@@ -195,3 +196,37 @@ click_visuallize_button_in_discover_page:
...
+    ref: kibana_expand_index_patterns
+  action:
+    if_element:
+      selector:
+        xpath: //span[@class="euiSideNavItemButton__label" and contains(., "Index Patterns")]
```
Comment (QiaolingTang) [inline lib/rules/web/admin_console/4.9/logging_kibana.xyaml:219]: Same comment as above

File: features/step_definitions/logging.rb (hunk 5 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1087]: To be more accurate, we need to check if the pod is printing logs. How about keep it ASIS and update it in future.

File: features/step_definitions/logging.rb (hunk 9 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1152]: @QiaolingTang  In this step, it only create clusterlogging instance  when the ES doesn't have persistent volume.  it assume we can create clusterlogging manually before run case.

File: lib/rules/web/admin_console/4.9/logging_kibana.xyaml [hunk 13]
```diff
@@ -195,3 +196,37 @@ click_visuallize_button_in_discover_page:
...
+search_doc_in_kibana:
+  elements:
+  - selector:
+      xpath: //input[@type="text" and @class="euiFieldText kuiLocalSearchAssistedInput__input euiFieldText--withIcon euiFieldText--fullWidth"]
+    op: clear
+  - selector:
+      xpath: //input[@type="text" and @class="euiFieldText kuiLocalSearchAssistedInput__input euiFieldText--withIcon euiFieldText--fullWidth"]
+    op: send_keys "<search_string>"
+  - selector:
+      xpath: //span[contains(@class, "euiButton__text euiSuperUpdateButton__text")]
```
Comment (QiaolingTang) [inline lib/rules/web/admin_console/4.9/logging_kibana.xyaml:231]: I suggest to use `xpath: //button[@data-test-subj="querySubmitButton" and @type="button"]`

File: features/step_definitions/logging.rb (hunk 5 above)
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1087]: Please update here.

File: features/step_definitions/logging.rb (hunk 8 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1138]: Why do you think that is dangerous? the cucushift can't handle while wisely?

Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1134]: No, why do you use `true` here? Using `true` makes me think it's an endless loop.

File: features/step_definitions/logging.rb (hunk 7 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1100]: @QiaolingTang   How about openshift-Logging admin who can modify the ES directly?

File: features/step_definitions/logging.rb (hunk 9 above)
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1148]: What's the usage of ` create clusterlogging manually before run case` ? I think we're doing some duplicated things here.

File: features/step_definitions/logging.rb (hunk 7 above)
Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1100]: How about `create index pattern via rest api`?

Comment (anpingli) [reply features/step_definitions/logging.rb:1100]: The step must be running as project admin.  we use cluster-admin here.  I think it is better to keep some works like 'cluster admin, logging admin'.   so we can know from the name that we must switch to cluster-admin before call this step.

Comment (anpingli) [reply features/step_definitions/logging.rb:1100]: Any keywords to make that clear if the step need the admin role or not?

Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1100]: > The step must be running as project admin. I think it is better to keep some works like 'cluster admin, logging admin'. so we can know from the name that we must switch to cluster-admin before call this step.

I see, thank you.
BTW, it switches to cluster-admin in this step, so no need to switch to cluster-admin before calling this step.
Besides, it seems this step is not used.

File: features/step_definitions/logging.rb (hunk 9 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1148]: This step can provide the ability to run upgrade on a given cluster. For example:  Cluster logging with customized configuration.

Comment (anpingli) [reply features/step_definitions/logging.rb:1148]: That is not duplicated with you step, that is an extend to your step.

File: features/step_definitions/logging.rb [hunk 14]
```diff
@@ -1056,3 +1056,120 @@
...
+    es_num ||= 3
+    redundancy_policy="SingleRedundancy"
+    deploy_cluster_logging=true
+
+    if cluster_logging('instance').exists?
+       deploy_cluster_logging=false
+       unless cluster_logging('instance').logstore_storage.has_key?("storageClassName")
+           deploy_cluster_logging=true
+       end
+    end
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1160]: Per my understanding, this block can be simplified to:
```
       if cluster_logging('instance').logstore_storage.has_key?("storageClassName")
           deploy_cluster_logging=false
       end
```

File: features/step_definitions/logging.rb (hunk 8 above)
Comment (anpingli) [reply features/step_definitions/logging.rb:1155]: Use ' while true' enable us to abort the loop raising message. ( raise("Can not find documents using clause #{lucene_query_string} under pattern  #{ pattern_name}! after 3 retry,  abort")
)

Comment (QiaolingTang) [reply features/step_definitions/logging.rb:1155]: I still think using endless loop is dangerous, if you want to raise error after 3 retries, I suggest to change the code to:
```
retries = 3
succeed=false
while retries > 0
  retries -= 1
  step %Q/I run the :check_log_count web action/
  if @result[:success]
    succeed=true
    break
  else
    log("Can not find documents using clause #{lucene_query_string} under pattern  #{ pattern_name}! retry.....")
  end
end
raise("Can not find documents using clause #{lucene_query_string} under pattern  #{ pattern_name}! after 3 retry,  abort") unless succeed
```

File: features/step_definitions/logging.rb (hunk 14 above)
Comment (anpingli) [reply features/step_definitions/logging.rb]: No. I got error below using step above


    Given I have clusterlogging with persistent storage ES                                                     # features/step_definitions/logging.rb:1169
      [06:36:23] INFO> Shell Commands: oc get clusterloggings instance --output=yaml --kubeconfig=/home/anli/workdir/laptop-anli-e/ocp4_admin.kubeconfig --namespace=logging-upg-prep-share
      
      STDERR:
      Error from server (NotFound): clusterloggings.logging.openshift.io "instance" not found
      
      [06:36:25] INFO> Exit Status: 1
      clusterloggings 'instance' not found (BushSlicer::ResourceNotFoundError)
      /home/anli/cucushift-runtime/verification-tests/lib/openshift/resource.rb:63:in `get_checked'
      /home/anli/cucushift-runtime/verification-tests/lib/openshift/resource.rb:127:in `get_cached_prop'
      /home/anli/cucushift-runtime/verification-tests/lib/openshift/resource.rb:134:in `raw_resource'
      /home/anli/cucushift-runtime/verification-tests/lib/openshift/cluster_logging.rb:132:in `log_store_spec'
      /home/anli/cucushift-runtime/verification-tests/lib/openshift/cluster_logging.rb:157:in `logstore_storage'
      /home/anli/cucushift-runtime/verification-tests/features/step_definitions/logging.rb:1183:in `/^I have clusterlogging with(?: (\d+))? persistent storage ES$/'
      features/upgrade/logging/upgrade.feature:16:in `I have clusterlogging with persistent storage ES'

Comment (QiaolingTang) [reply features/step_definitions/logging.rb]: I saw `--namespace=logging-upg-prep-share`, be careful with the namespace

File: features/step_definitions/logging.rb [hunk 15]
```diff
@@ -1056,3 +1056,142 @@
...
+
+Given /^I have clusterlogging with(?: (\d+))? persistent storage ES$/ do |es_num|
+    # I switch to cluster admin pseudo user
+    ensure_admin_tagged
+    es_num ||= 3
+    redundancy_policy="SingleRedundancy"
+    deploy_cluster_logging=true
+
+    if project('openshift-logging').exists?
+        step %Q/I use the "openshift-operators-redhat" project/
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1177]: @anpingli `openshift-operators-redhat` -- > `openshift-logging`

File: features/step_definitions/logging.rb [hunk 16]
```diff
@@ -1056,3 +1056,142 @@
...
+    deploy_pod = true
+    log_type ||= "any"
+    case log_type
+    when "json"
+      template_file = "#{file_dir}/container_json_log_template.json"
+    when "unicode", "flat"
+      template_file = "#{file_dir}/container_json_unicode_log_template.json"
+    else
+      template_file = "#{file_dir}/container_json_log_template.json"
+    end
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:1071]:
```
template_file = "#{file_dir}/container_json_log_template.json"
if log_type == "unicode" || log_type == "flat"
  template_file = "#{file_dir}/container_json_unicode_log_template.json"
end
```

**Comment by anpingli** [comment]
The regression result: 
upgrade case:  OCP-22911 https://mastern-jenkins-csb-openshift-qe.apps.ocp4.prod.psi.redhat.com/job/ocp-common/job/Runner/178112/console
kibana cases:  OCP-30362,OCP-30361 https://mastern-jenkins-csb-openshift-qe.apps.ocp4.prod.psi.redhat.com/job/ocp-common/job/Runner/178149/console

**Comment by anpingli** [comment]
@pruan-rht  Could you help me review this PR?

File: features/step_definitions/logging.rb [hunk 17]
```diff
@@ -1056,3 +1056,139 @@
...
+    rest_cmd="curl -s --connect-timeout 10  -XPOST \"http://localhost:5601/api/saved_objects/index-pattern/#{pattern_name}\" -H \"kbn-xsrf: true\" -H \"x-forwarded-user: #{user_name}\" -H \"securitytenant: __user__\" -H \"Authorization: Bearer #{user_token}\" -H \"Content-Type: application/json\" -d \'{\"attributes\": {\"title\": \"#{pattern_name}\", \"timeFieldName\": \"@timestamp\"}}\'"
+
+    step %Q/a pod becomes ready with labels:/, table(%{
+      | component= kibana|
+    })
+    @result = pod.exec("bash", "-c", rest_cmd, as: admin, container: 'kibana')
+    raise "Failed to create index pattern via rest api" unless @result[:success]
+end
+
+Given /^I can display the(?: "(.+)")? pod logs of the(?: "(.+)")? project under the(?: "(.+)")? pattern in kibana$/ do | pod_name, project_name, pattern_name |
```
Comment (pruan-rht) [inline features/step_definitions/logging.rb:1134]: @anpingli if you looks at [_constants.rb](https://github.com/openshift/verification-tests/blob/master/features/step_definitions/_constants.rb) file, there contains some useful alias that you can use in the step definition.

File: features/step_definitions/logging.rb [hunk 18]
```diff
@@ -1056,3 +1056,139 @@
...
+    end
+
+    if deploy_cluster_logging
+      if(es_num==1)
+        redundancy_policy="ZeroRedundancy"
+      end 
+      step %Q/default storageclass is stored in the :default_sc clipboard/
+      step %Q|I obtain test data file "logging/clusterlogging/clusterlogging-storage-template.yaml"|
+      step %Q/I create clusterlogging instance with:/, table(%{
+        | crd_yaml            | clusterlogging-storage-template.yaml |
```
Comment (pruan-rht) [inline features/step_definitions/logging.rb:1187]: @anpingli table alignment `|`

File: features/upgrade/logging/upgrade.feature [hunk 19]
```diff
@@ -6,88 +6,78 @@ Feature: Logging upgrading related features
...
-      | storage_size        | 20Gi                                 |
-      | es_node_count       | 3                                    |
-      | redundancy_policy   | SingleRedundancy                     |
-    Then the step should succeed
-    Given I wait for the project "logging-upgrade-data-check" logs to appear in the ES pod
-    # check cron jobs
+    Given I have "json" log pod in project "logging-upg-prep-1"
+    And I have "json" log pod in project "logging-upg-prep-share"
+    Given I run the :oadm_groups_new admin command with:
+      | group_name | project-group-share |
```
Comment (pruan-rht) [inline features/upgrade/logging/upgrade.feature:13]: @anpingli table alignment `|`

File: features/upgrade/logging/upgrade.feature [hunk 20]
```diff
@@ -6,88 +6,78 @@ Feature: Logging upgrading related features
...
     When I check the cronjob status
     Then the step should succeed
 
     Given I switch to the first user
     When I login to kibana logging web console
+    Given I have index pattern "*app"
+    Then I can display the pod logs of the "logging-upg-prep-1" project under the "*app" pattern in kibana
+    Then I close the current browser
+    Given I run the :policy_add_role_to_group client command with:
+      | group_name | project-group-share |
```
Comment (pruan-rht) [inline features/upgrade/logging/upgrade.feature:27]: @anpingli table alignment

**Review by pruan-rht — COMMENTED** [review]
/lgtm

**Pull request merged**